store.go

```diff
@@ -369,3 +377,71 @@ func (c *controller) processEndpointUpdate(ep *endpoint) bool {
 
 	return false
 }
+
+func (c *controller) validateLocalDatastoreConfig() bool {
+	return c.cfg != nil && c.cfg.LocalDataStore.Client.Provider != "" && c.cfg.LocalDataStore.Client.Address != ""
+}
+
+func (c *controller) getNetworksFromLocalStore() ([]*store.KVPair, error) {
+	c.Lock()
+	cs := c.localStore
+	c.Unlock()
+	return cs.KVStore().List(datastore.Key(datastore.NetworkKeyPrefix))
+}
+
+func (c *controller) getDataStore(scope driverapi.Scope) (datastore datastore.DataStore) {
+	c.Lock()
+	if scope == driverapi.GlobalScope {
+		datastore = c.store
+	} else if scope == driverapi.LocalScope {
+		datastore = c.localStore
+	}
+	c.Unlock()
+	return
+}
+
+func (c *controller) getEndpointsFromLocalStore() ([]*store.KVPair, error) {
+	c.Lock()
+	cs := c.localStore
+	c.Unlock()
+	return cs.KVStore().List(datastore.Key(datastore.EndpointKeyPrefix))
+}
+
+func (c *controller) processEndpointsUpdate(eps []*store.KVPair, prune *endpointTable) {
+	for _, epe := range eps {
+		var ep endpoint
+		err := json.Unmarshal(epe.Value, &ep)
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		if prune != nil {
+			delete(*prune, ep.id)
+		}
+		ep.SetIndex(epe.LastIndex)
+		if nid, err := networkIDFromKey(epe.Key); err != nil {
+			log.Error(err)
+			continue
+		} else {
+			if n, err := c.NetworkByID(nid); err != nil {
+				log.Error(err)
+				continue
+			} else {
+				ep.network = n.(*network)
+			}
+		}
+		if c.processEndpointUpdate(&ep) {
+			err = c.newEndpointFromStore(epe.Key, &ep)
+			if err != nil {
+				log.Error(err)
+			}
+		}
+	}
+}
+
+func networkIDFromKey(key string) (string, error) {
+	if !strings.HasPrefix(key, "docker/libnetwork/endpoint/") {
```

Comment:
Lets not expose the key structure outside of the object that manages the key.
For example, the key structure for endpoint must be kept as part of endpoint.go.
Hence please remove this function and reuse the `networkIDFromKey` in endpoint.go and ofcourse make it a public function.
Also, `networkIDFromKey` in endpoint.go is unused now, hence you can even change the signature to pass a key string instead of the string[] as it is currently defined.